internal/controller/etcdopstask/reconcile_task.go

```diff
@@ -0,0 +1,222 @@
+// SPDX-FileCopyrightText: 2025 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package etcdopstask
+
+import (
+	"context"
+	"fmt"
+
+	druidapicommon "github.com/gardener/etcd-druid/api/common"
+	druidv1alpha1 "github.com/gardener/etcd-druid/api/core/v1alpha1"
+	"github.com/gardener/etcd-druid/internal/controller/etcdopstask/handler"
+	ctrlutils "github.com/gardener/etcd-druid/internal/controller/utils"
+	druiderr "github.com/gardener/etcd-druid/internal/errors"
+	"github.com/gardener/etcd-druid/internal/utils/kubernetes"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	// ErrDuplicateTask represents the error in case of a duplicate task for the same etcd.
+	ErrDuplicateTask druidv1alpha1.ErrorCode = "ERR_DUPLICATE_TASK"
+)
+
+// reconcileTask manages the lifecycle of an EtcdOpsTask resource.
+// It executes a series of step functions to ensure the task is processed correctly.
+func (r *Reconciler) reconcileTask(ctx context.Context, logger logr.Logger, taskObjKey client.ObjectKey, taskHandler handler.Handler) ctrlutils.ReconcileStepResult {
+	logger = logger.WithValues("op", "reconcileTask")
+	logger.Info("Triggering task execution flow")
+
+	reconcileStepFns := []reconcileFn{
+		r.ensureTaskFinalizer,
+		r.transitionToPendingState,
+		r.admitTask,
+		r.transitionToInProgressState,
+		r.runTask,
+	}
+
+	for _, fn := range reconcileStepFns {
+		if result := fn(ctx, logger, taskObjKey, taskHandler); ctrlutils.ShortCircuitReconcileFlow(result) {
+			return result
+		}
+	}
+	logger.Info("Task execution flow completed")
+	return ctrlutils.ContinueReconcile()
+}
+
+// ensureTaskFinalizer checks if the EtcdOpsTask has the finalizer.
+// If not, it adds the finalizer and updates the task status.
+func (r *Reconciler) ensureTaskFinalizer(ctx context.Context, _ logr.Logger, taskObjKey client.ObjectKey, _ handler.Handler) ctrlutils.ReconcileStepResult {
+	task, err := r.getTask(ctx, taskObjKey)
+	if err != nil {
+		return ctrlutils.ReconcileWithError(err)
+	}
+
+	if controllerutil.ContainsFinalizer(task, druidapicommon.EtcdOpsTaskFinalizerName) {
+		return ctrlutils.ContinueReconcile()
+	}
+
+	if err := kubernetes.AddFinalizers(ctx, r.client, task, druidapicommon.EtcdOpsTaskFinalizerName); err != nil {
```

Comment:
no need to short assign err. Take your IDE's help to figure out where you should not short assign. There are a lot of places in the new code (introduced as part of this PR) where the same problem is present.